### PR TITLE
Easier package mgr overrides

### PR DESCRIPTION
**Pull request by bcoca**
##### ISSUE TYPE

- Feature Pull Request

### CODE REVIEW

File: lib/ansible/plugins/action/package.py [hunk 1]
```diff
@@ -45,14 +46,20 @@ def run(self, tmp=None, task_vars=None):
...
-                    module = self._templar.template("{{hostvars['%s']['ansible_facts']['pkg_mgr']}}" % self._task.delegate_to)
+                    module = C.config.get_config_value('PACKAGE_MANAGER_OVERRIDE', variables=combine_vars(self._task.vars, task_vars.get('delegated_vars', {})))
+                    if module is None:
+                        dh = '%s' % self._task.delegate_to
+                        module = self._templar.template(f"{{hostvars['{dh}']['ansible_local']['pkg_mgr']|default(hostvars['{dh}']['ansible_facts']['pkg_mgr']})}}"
                 else:
-                    module = self._templar.template('{{ansible_facts.pkg_mgr}}')
+                    module = C.config.get_config_value('PACKAGE_MANAGER_OVERRIDE', variables=task_vars)
+                    if module is None:
+                        module = self._templar.template('{{ansible_local.pkg_mgr|default(ansible_facts.pkg_mgr)}}')
```
Comment (sivel) [inline lib/ansible/plugins/action/package.py]: Something I forgot about `ansible_local`, is that is namespaces things by filename.  So we might have to do something like a file named `pkg_mgr.fact` with contents of `{"value": "foo"}` and use `ansible_local.pkg_mgr.value`

File: lib/ansible/modules/package.py [hunk 2]
```diff
@@ -40,6 +40,8 @@
     description:
       - The required package manager module to use (V(dnf), V(apt), and so on). The default V(auto) will use existing facts or try to autodetect it.
       - You should only use this field if the automatic selection is not working for some reason.
+      - Since version 2.17 you can override the automatic detection both via a local facts by setting 'pkg_mgr' or override both by using the
```
Comment (samccann) [inline lib/ansible/modules/package.py]:
```suggestion
      - Since version 2.17 you can override the automatic detection both through a local facts by setting ``pkg_mgr`` or override both by using the
```

File: lib/ansible/modules/package.py [hunk 3]
```diff
@@ -40,6 +40,8 @@
...
+        PACKAGE_MANAGER_OVERRIDE configuration settable via ANSIBLE_PACKAGE_MANAGER_OVERRIDE enviornment variable or ansible_package_manager_override variable.
```
Comment (samccann) [inline lib/ansible/modules/package.py]:
```suggestion
        PACKAGE_MANAGER_OVERRIDE configuration settable with the ANSIBLE_PACKAGE_MANAGER_OVERRIDE enviornment variable or the ``ansible_package_manager_override`` variable.
```

**Review by samccann — COMMENTED** [review]
a little 'just say no to Latin' nits :-)

**Comment by maxamillion** [comment]
Idle curiosity. With this change, is it possible for a playbook or role to use `set_fact` to override the `pkg_mgr` mid-flight during a playbook?

**Comment by sivel** [comment]
> Idle curiosity. With this change, is it possible for a playbook or role to use `set_fact` to override the `pkg_mgr` mid-flight during a playbook?

Technically yes.  But this introduces a never before utilized variable, meant to define a package manager override for a system. It could be abused, but should offer an easy way to override, or define for a host.  Even by just dropping a file on the remote host, and not defining it within the playbook.

**Comment by maxamillion** [comment]
@sivel +1 - thank you

File: lib/ansible/plugins/action/package.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/plugins/action/package.py]: or name it override? can be expanded to more things, like service_mgr

File: lib/ansible/modules/package.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/modules/package.py]: @samccann as a bilingual spanish speaker that is not proficient in either language, its really hard to me sometimes to even realize it is a latinism

File: lib/ansible/plugins/action/package.py [hunk 4]
```diff
@@ -45,14 +46,20 @@ def run(self, tmp=None, task_vars=None):
         if module == 'auto':
             try:
                 if self._task.delegate_to:  # if we delegate, we should use delegated host's facts
-                    module = self._templar.template("{{hostvars['%s']['ansible_facts']['pkg_mgr']}}" % self._task.delegate_to)
+                    module = C.config.get_config_value('PACKAGE_MANAGER_OVERRIDE', variables=combine_vars(self._task.vars, task_vars.get('delegated_vars', {})))
+                    if module is None:
+                        dh = '%s' % self._task.delegate_to
+                        module = self._templar.template(f"{{hostvars['{dh}']['ansible_local']['overrides']]['pkg_mgr']|default(hostvars['{dh}']['ansible_facts']['pkg_mgr'])}}"
```
Comment (webknjaz) [inline lib/ansible/plugins/action/package.py]:
```suggestion
                        module = self._templar.template(f"{{hostvars['{dh}']['ansible_local']['overrides']]['pkg_mgr']|default(hostvars['{dh}']['ansible_facts']['pkg_mgr'])}}")
```

File: lib/ansible/plugins/action/package.py [hunk 5]
```diff
@@ -45,14 +47,20 @@ def run(self, tmp=None, task_vars=None):
         if module == 'auto':
             try:
                 if self._task.delegate_to:  # if we delegate, we should use delegated host's facts
-                    module = self._templar.template("{{hostvars['%s']['ansible_facts']['pkg_mgr']}}" % self._task.delegate_to)
+                    module = C.config.get_config_value('PACKAGE_MANAGER_OVERRIDE', variables=combine_vars(self._task.vars, task_vars.get('delegated_vars', {})))
+                    if module is None:
+                        dh = '%s' % self._task.delegate_to
+                        module = self._templar.template(f"{{hostvars['{dh}']['ansible_local']['overrides']]['pkg_mgr']|default(hostvars['{dh}']['ansible_facts']['pkg_mgr'])}}")
```
Comment (sivel) [inline lib/ansible/plugins/action/package.py]:
```suggestion
                        module = self._templar.template(
                            f"{{hostvars['{dh}']['ansible_local']['overrides']]['pkg_mgr']|default(hostvars['{dh}']['ansible_facts']['pkg_mgr'])}}"
                        )
```

**Comment by sivel** [comment]
fwiw, we can merge this ahead of my [PR](https://github.com/ansible/ansible/pull/82881) to address how the templating happens in package to safely evaluate the expression, specifically for `delegate_to`.  The plan for my PR is to merge after stable-2.17 branching but before any releases.  Ultimately the change here is needed for DT also, so it probably needs to merge to devel, but will conflict with the DT changes already in this file.

**Comment by sivel** [comment]
We should probably add tests for this as well 🙂

File: lib/ansible/modules/package.py [hunk 6]
```diff
@@ -40,6 +40,8 @@
     description:
       - The required package manager module to use (V(dnf), V(apt), and so on). The default V(auto) will use existing facts or try to autodetect it.
       - You should only use this field if the automatic selection is not working for some reason.
+      - Since version 2.17 you can override the automatic detection both through a local facts C(overrides) file and setting
```
Comment (nitzmahone) [inline lib/ansible/modules/package.py]: We've been trying to avoid embedding version numbers in prose like this- docs are versioned, as is the option, so we can just describe the feature as it exists.

File: lib/ansible/plugins/action/package.py [hunk 7]
```diff
@@ -45,14 +47,22 @@ def run(self, tmp=None, task_vars=None):
         if module == 'auto':
             try:
                 if self._task.delegate_to:  # if we delegate, we should use delegated host's facts
-                    module = self._templar.template("{{hostvars['%s']['ansible_facts']['pkg_mgr']}}" % self._task.delegate_to)
+                    module = C.config.get_config_value('PACKAGE_MANAGER_OVERRIDE', variables=combine_vars(self._task.vars, task_vars.get('delegated_vars', {})))
```
Comment (nitzmahone) [inline lib/ansible/plugins/action/package.py]: This doesn't look like it handles the case where the override is itself a template.

File: lib/ansible/plugins/action/package.py [hunk 8]
```diff
@@ -45,14 +47,22 @@ def run(self, tmp=None, task_vars=None):
...
+                    if module is None:
+                        dh = '%s' % self._task.delegate_to
+                        module = self._templar.template(
+                            f"{{hostvars['{dh}']['ansible_local']['overrides']]['pkg_mgr']|default(hostvars['{dh}']['ansible_facts']['pkg_mgr'])}}"
+                        )
                 else:
-                    module = self._templar.template('{{ansible_facts.pkg_mgr}}')
+                    module = C.config.get_config_value('PACKAGE_MANAGER_OVERRIDE', variables=task_vars)
```
Comment (nitzmahone) [inline lib/ansible/plugins/action/package.py]: Ditto as above- what if it's a template?

File: lib/ansible/plugins/action/package.py [hunk 9]
```diff
@@ -45,14 +47,22 @@ def run(self, tmp=None, task_vars=None):
         if module == 'auto':
             try:
                 if self._task.delegate_to:  # if we delegate, we should use delegated host's facts
-                    module = self._templar.template("{{hostvars['%s']['ansible_facts']['pkg_mgr']}}" % self._task.delegate_to)
+                    module = C.config.get_config_value('PACKAGE_MANAGER_OVERRIDE', variables=combine_vars(self._task.vars, task_vars.get('delegated_vars', {})))
+                    if module is None:
+                        dh = '%s' % self._task.delegate_to
+                        module = self._templar.template(
+                            f"{{hostvars['{dh}']['ansible_local']['overrides']]['pkg_mgr']|default(hostvars['{dh}']['ansible_facts']['pkg_mgr'])}}"
```
Comment (nitzmahone) [inline lib/ansible/plugins/action/package.py]: Eek, template injection... DT has `Templar.evaluate_expression()` for this, so you wouldn't need to wrap it in `{{ }}`, and it supports passing template locals so you wouldn't need to meta-template the expression first.

File: lib/ansible/plugins/action/package.py [hunk 10]
```diff
@@ -45,14 +47,22 @@ def run(self, tmp=None, task_vars=None):
...
                 else:
-                    module = self._templar.template('{{ansible_facts.pkg_mgr}}')
+                    module = C.config.get_config_value('PACKAGE_MANAGER_OVERRIDE', variables=task_vars)
+                    if module is None:
+                        module = self._templar.template("{{ansible_local['overrides']['pkg_mgr']|default(ansible_facts['pkg_mgr'])}}")
             except Exception:
                 pass  # could not get it from template!
 
         try:
             if module == 'auto':
```
Comment (nitzmahone) [inline lib/ansible/plugins/action/package.py:46]: Seems like this can never run now- won't the `auto` value be trashed with a `None` by the config calls above?

File: lib/ansible/plugins/action/package.py [hunk 11]
```diff
@@ -45,14 +47,22 @@ def run(self, tmp=None, task_vars=None):
         if module == 'auto':
             try:
                 if self._task.delegate_to:  # if we delegate, we should use delegated host's facts
```
Comment (nitzmahone) [inline lib/ansible/plugins/action/package.py]: Assuming the fact run is working, doesn't doing the override logic first preclude a custom fact module from setting this? Maybe that's a feature, not a bug, but we should probably be explicit about that (and enforce it with a test?) if it's the desired behavior.

**Review by nitzmahone — CHANGES_REQUESTED** [review]
A few potential holes to plug, and of course, everyone's favorite statement: "this would be a lot easier under DT". :laughing:

File: lib/ansible/modules/package.py (hunk 6 above)
Comment (bcoca) [reply lib/ansible/modules/package.py]: We normally add it to the prose if it modifies an existing option so users reading the 'latest' do not confuse it with it working on older versions. If this added a new option with `version_added`, it would not be in the prose.

File: lib/ansible/plugins/action/package.py (hunk 9 above)
Comment (sivel) [reply lib/ansible/plugins/action/package.py]: Yeah, I'll take care of this in stable-2.17 branch after this is merged.  It'll be extended in https://github.com/ansible/ansible/pull/82881

Comment (bcoca) [reply lib/ansible/plugins/action/package.py]: we can update then when DT hits, but for 2.17 and backports ...

File: lib/ansible/plugins/action/package.py (hunk 8 above)
Comment (bcoca) [reply lib/ansible/plugins/action/package.py]: ~config system can template if it has variables, which we pass in~ only for defaults, we don't template any other config so we should keep this the same.

File: lib/ansible/plugins/action/package.py (hunk 11 above)
Comment (bcoca) [reply lib/ansible/plugins/action/package.py]: i specified now in docs how it works, but our own 'fact finding' still uses setup.py directly  .. .not sure i want to use gather_facts as it is not guaranteed we'll get pgk_mgr info if they customized it, nor do i want to make it x2 expensive and fallback to setup .. but could ...

File: lib/ansible/plugins/action/package.py (hunk 9 above)
Comment (bcoca) [reply lib/ansible/plugins/action/package.py]: I went back and realized that for configs we only template defaults, so removed the templating in this case, leaving  note for @sivel's PR to introduce delegate_to templating.

File: lib/ansible/plugins/action/package.py (hunk 7 above)
Comment (bcoca) [reply lib/ansible/plugins/action/package.py]: nope, and it shouldn't

**Comment by sivel** [comment]
@nitzmahone and I just discussed some of the implementation goals and architecture around this.  Instead of in line review comments, I just leave an update about what we would like to see.  I don't think there will be much argument about it 🙂 

We'd like to limit the scope a bit more.  We may have gotten a little too excited with trying to find a use for `ansible_local`.

Overall, let's just move forward with a single new var for now, and chop off the ini and env vars.  We both feel like because it's so easy to add them, we're adding too many ini and env vars for things that just make more sense as hostvars.  If it needs to be global, `-e` or `group_vars/all`.

For the time being lets just add `ansible_package_manager_override` and drop the rest.  There may be some things coming that data tagging gives us, or maybe some additional changes we can make when we have more time so we could implement some functionality where facts won't necessarily get blown away.  We'd need to think about it more.  Ideally I think it would have been great if we could just leverage `ansible_pkg_mgr` in this way, but we have no facility. So to avoid overloading, let's stick with just the 1.

So to recap:

* drop ini/env configuration
* drop ansible_local override

**Comment by bcoca** [comment]
One of the requirements was being able to set this on the EE, which they normally do by env vars, if they have to shift to alter inventory it won't work 'seamless' to the users as they upgrade the EE.

File: lib/ansible/plugins/action/package.py [hunk 12]
```diff
@@ -38,31 +40,47 @@ def run(self, tmp=None, task_vars=None):
...
-                    module_name='ansible.legacy.setup',
-                    module_args=dict(filter='ansible_pkg_mgr', gather_subset='!all'),
-                    task_vars=task_vars)
-                display.debug("Facts %s" % facts)
-                module = facts.get('ansible_facts', {}).get('ansible_pkg_mgr', 'auto')
-
-            if module != 'auto':
+
+                if self._task.delegate_to:
+                    # TODO: template delegate_to if needed
```
Comment (sivel) [inline lib/ansible/plugins/action/package.py]: We shouldn't need to worry with `delegate_to` not being a template, the TE should already be handling this as needed.

Comment (bcoca) [reply lib/ansible/plugins/action/package.py]: i might be misremembering, but i thought that was the issue you were fixing in another pr

File: lib/ansible/plugins/action/package.py [hunk 13]
```diff
@@ -38,31 +39,46 @@ def run(self, tmp=None, task_vars=None):
...
+                else:
+                    hosts_vars = task_vars
+                    tvars = task_vars
+
+                # use config
+                module = tvars.get('ansible_package_use', None)
+
+                if not module:
+                    # no use, no config, get from facts
+                    if hosts_vars.get('_ansible_facts_gathered', False):
```
Comment (sivel) [inline lib/ansible/plugins/action/package.py]: I don't think this will be sufficient.  It's set by `gather_facts` but there is nothing saying that `gather_subset` may have excluded `pkg_mgr`.

`gather_subset: ['!min']` for example would prevent it from being gathered.

**Review by sivel — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/action/package.py (hunk 13 above)
Comment (sivel) [reply lib/ansible/plugins/action/package.py]: Going back to `try/except KeyError` may be best

Comment (bcoca) [reply lib/ansible/plugins/action/package.py]: made it conditional on the actual data

**Review by sivel — APPROVED** [review]

**Review by nitzmahone — APPROVED** [review]
